Respect TorchDispatchMode for shallow_copy_and_detach

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83330
* #83380
* __->__ #83372

I noticed I was missing tensor creations with modes when I tried
to delete proxy tensor.  This was the cause.

Hypothetically, all PyInterpreter calls could get this treatment.
But I think it only matters for detach; the rest do not return
Tensors and most modes will not be interested in them.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Respect TorchDispatchMode for shallow_copy_and_detach

I noticed I was missing tensor creations with modes when I tried
to delete proxy tensor.  This was the cause.

Hypothetically, all PyInterpreter calls could get this treatment.
But I think it only matters for detach; the rest do not return
Tensors and most modes will not be interested in them.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Respect TorchDispatchMode for shallow_copy_and_detach"


I noticed I was missing tensor creations with modes when I tried
to delete proxy tensor.  This was the cause.

Hypothetically, all PyInterpreter calls could get this treatment.
But I think it only matters for detach; the rest do not return
Tensors and most modes will not be interested in them.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Respect TorchDispatchMode for shallow_copy_and_detach"


I noticed I was missing tensor creations with modes when I tried
to delete proxy tensor.  This was the cause.

Hypothetically, all PyInterpreter calls could get this treatment.
But I think it only matters for detach; the rest do not return
Tensors and most modes will not be interested in them.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Respect TorchDispatchMode for shallow_copy_and_detach"


I noticed I was missing tensor creations with modes when I tried
to delete proxy tensor.  This was the cause.

Hypothetically, all PyInterpreter calls could get this treatment.
But I think it only matters for detach; the rest do not return
Tensors and most modes will not be interested in them.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

zou3519 commented on test/test_python_dispatch.py:
```diff
@@ -872,6 +872,24 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        seen = set()
+        test_case = self
+
+        class TestMode(TorchDispatchMode):
+            def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+                tree_map_only(torch.Tensor, lambda t: test_case.assertIn(t, seen), (args, kwargs))
+                if kwargs is None:
+                    kwargs = {}
+                r = func(*args, **kwargs)
+                tree_map_only(torch.Tensor, lambda t: seen.add(t), r)
+                return r
```
This is pretty cool

zou3519 commented on c10/core/TensorImpl.cpp:
```diff
@@ -537,17 +538,24 @@ template <typename VariableVersion>
+  c10::intrusive_ptr<TensorImpl> r;
+  const auto& maybe_torch_dispatch_mode_state =
+      c10::impl::TorchDispatchModeTLS::get_state();
+  // TODO: do we have to exclude after Python dispatch key set?
```
What do you mean by "exclude after Python dispatch key set"?

ezyang replied:
I noticed that in PythonFallback, before calling into Python we exclude all dispatch keys after the Python key. I'm not sure why this is being done (@albanD knows maybe?) but I haven't replicated that logic here.

albanD replied:
Ho that sounds wrong, it should be excluding all *before* the Python key IIUC

ezyang replied:
here's the relevant code

```
void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
  TORCH_INTERNAL_ASSERT(tls_on_entry.has_value());
  // c10::impl::ForceDispatchKeyGuard dispatcher_guard(tls_on_entry.value());
  // StashTLSOnEntryGuard stash_guard;
  c10::impl::ExcludeDispatchKeyGuard guard(after_Python_keyset);
```

bdhirsh commented on c10/core/TensorImpl.cpp:
```diff
@@ -537,17 +538,24 @@ template <typename VariableVersion>
+  if (r) {
+    r->set_version_counter(std::forward<VariableVersion>(version_counter));
+    r->set_allow_tensor_metadata_change(allow_tensor_metadata_change);
+    return r;
```
I think we're going to need to make a similar change for the other tensor subclasses that override shallow_copy_and_detach().

`FunctionalTensorWrapper` ([here](https://github.com/pytorch/pytorch/blob/d4bd88b64be3d7156cf8617e33faae8eca307ce4/aten/src/ATen/FunctionalTensorWrapper.cpp#L300))

(more recently) `NestedTensorImpl` ([here](https://github.com/pytorch/pytorch/blob/d4bd88b64be3d7156cf8617e33faae8eca307ce4/aten/src/ATen/NestedTensorImpl.cpp#L249))

Although coming up with a better pattern to avoid the code dup'ing would probably be better longterm.

zou3519 replied:
I feel like we've been tiptoe-ing around the actual issue, which is that shallow_copy_and_detach is not a dispatch-able operation. If it were, then everything would just work out. This has led to issues in the past, like https://github.com/pytorch/pytorch/issues/71725 and https://github.com/pytorch/pytorch/issues/74991 (which is exactly what you're referring to, @bdhirsh).

ezyang replied:
BTW, having the tensor subclasses properly wrap this will only be necessary when we want to directly trace operations involving those tensors. NestedTensorImpl will need it, but FunctionalTensorWrapper will not (we always trace "below" functional tensor wrapper).

bdhirsh replied:
> NestedTensorImpl will need it, but FunctionalTensorWrapper will not (we always trace "below" functional tensor wrapper)

I think FunctionalTensorWrapper will for LTC/XLA (although to be fair, that's assuming that the way XLA integrates with AOTAutograd-style tracing is using lazy tensor infra)

bdhirsh replied:
> This has led to issues in the past, like https://github.com/pytorch/pytorch/issues/71725 and https://github.com/pytorch/pytorch/issues/74991 (which is exactly what you're referring to, @bdhirsh). This probably won't happen soon, but do you think it would resolve the long-term problem?

Those gh issues were what I had in my head when making that comment :) Agreed they would fix this issue.

ezyang replied:
It doesn't feel like, to me, we will be able to get rid of `shallow_copy_and_detach` as a concept on TensorImpl though; the proposed fix would simply to make sure you only ever got here via a dispatching operation.

bdhirsh replied:
I was thinking more that that would give us a place to shunt some of this boilerplate code so it can be re-used (so at least, we can only have one copy of the code that knows to call torch_dispatch for all the different subclasses).

ezyang replied:
Yes. The point, though, is that `shallow_copy_and_detach` would remain, but it would not be dispatched. (And the common code goes in detach)

ezyang committed:
Update on "Respect TorchDispatchMode for shallow_copy_and_detach"


I noticed I was missing tensor creations with modes when I tried
to delete proxy tensor.  This was the cause.

Hypothetically, all PyInterpreter calls could get this treatment.
But I think it only matters for detach; the rest do not return
Tensors and most modes will not be interested in them.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]
